pkg/apis/logicalcluster/v1alpha1/cluster-provider-desc.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2022 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ClusterProviderType string
+
+const (
+	KindProviderType ClusterProviderType = "kind"
+)
+
+// ClusterProviderDesc represents a provider.
+//
+// +crd
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:resource:scope=Cluster,shortName=cpc
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ClusterProviderDesc struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// `spec` describes a provider.
+	// +optional
+	Spec ClusterProviderDescSpec `json:"spec,omitempty"`
+
+	// `status` describes the status of the cluster object.
+	// +optional
+	Status ClusterProviderDescStatus `json:"status"`
+}
+
+// ClusterProviderDescSpec describes a cluster.
+// TODO: We are currently only listing the type and config.
+// There will be additional fields in the future.
+type ClusterProviderDescSpec struct {
+	// ProviderType is the type of the provider of the cluster.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="cluster is immutable"
+	ProviderType ClusterProviderType `json:"ProviderType"`
+
+	// TODO: this should be stored as a secret!
+	// Config is the provider config
+	// +kubebuilder
+	Config string `json:"Config,omitempty"`
+
+	// ClusterPrefixForDiscovery contains the prefix used during cluster discovery.
```

Comment:
How exactly does the provider use this prefix?

Is this properly part of the generic provider config?  Couldn't different providers have different ways of doing (whatever) besides prefixing?

Reply:
The prefix just try to limit the scope of the discovered clusters. It allows the provider a simple way to distinguish the clusters that he wants to be discovered and other clusters that might be seen through the same provider endpoint but shouldn't be discovered/imported. It's not mandatory and if empty all clusters will be discovered. In container runtime PoC they had something similar (although hard coded) in which they only discovered `fleet-xxx`  clusters

Reply:
Yes, This is stage one in which the provider behavior is controlled through definitions on the cluster-manager side, so we came up with some common behavior. For phase 2,  we plan to support plugin and configuration of external providers not through the cluster manager. The provider client is implemented by the provider and can implement whatever technique they want for selection/filtering of the discovered clusters.

Reply:
@MikeSpreitzer I opened a new issue to address logical-clusters stage 2.  The above will be part of that.